txnsync/encodedgroupsunmarshalers.go

```diff
@@ -0,0 +1,925 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/transactions"
+)
+
+var errDataMissing = errors.New("failed to decode: data missing")
+
+func getSlice(b []byte, index int, size int) ([]byte, error) {
+	if index*size+size > len(b) {
+		return nil, errDataMissing
+	}
+	return b[index*size : index*size+size], nil
+}
+
+func getNibble(b []byte, index int) (byte, error) {
+	if index > len(b)*2 {
+		return 0, errDataMissing
+	}
+	if index%2 == 0 {
+		return b[index/2] / 16, nil
+	}
+	return b[index/2] % 16, nil
+}
+
+func (stub *txGroupsEncodingStub) reconstructSignedTransactions(signedTxns []transactions.SignedTxn, genesisID string, genesisHash crypto.Digest) (err error) {
+	var index int
+	index = 0
+	err = stub.BitmaskSig.Iterate(int(stub.TotalTransactionsCount), func(i int) error {
+		slice, err := getSlice(stub.Sig, index, len(crypto.Signature{}))
+		if err != nil {
+			return err
+		}
+		copy(signedTxns[i].Sig[:], slice)
+		index++
```

Comment:
The pattern that you've been using here is abit tricky ( but it is correct ).
I think that avoiding having the index and the slice would simplify the implementation and would be
more maintainable :

```golang
// inside the anonymous function
err := nextSlice(&stub.Sig, signedTxns[i].Sig[:], len(crypto.Signature{})
if err != nil {
    return err
}
// the nextSlice definition - copy the next slice and slide the src window.
func nextSlice(src *[]byte, dst []byte, size int) ([]byte, error) {
    if len(*src) < size {
         return errDataMissing
    }
    copy(dst[:], (*src)[: size])
    // slice the src window so next call would get the next entry.
    *src = (*src)[size:]
    return nil
}
```